File: Atividade 3/MergeSort.java

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.nio.file.Files;
import java.nio.file.Paths;

public class MergeSort {
    public static void main(String[] args) throws IOException {
        // Lê os números do arquivo de entrada
        List<Integer> numbers = readNumbers("D:\\Faculdade\\Analise\\An-lise-de-Desempenho\\Atividade 3\\arq (1).txt");

        // Inicia a medição de tempo e memória
        long startTime = System.currentTimeMillis();
        long startMemory = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();

        // Ordena os números
        List<Integer> sortedNumbers = mergeSort(numbers);

        // Termina a medição de tempo e memória
        long endTime = System.currentTimeMillis();
        long endMemory = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();

        // Escreve os números ordenados no arquivo de saída
        writeNumbers("arq-saidaMergeSort_Java.txt", sortedNumbers);

        // Imprime o tempo de execução e a memória utilizada
        System.out.printf("Tempo de execução: %d ms\n", (endTime - startTime));
        System.out.printf("Memória utilizada: %d KB\n", (endMemory - startMemory) / 1024);
    }

    // Função para ordenar uma lista de números usando Merge Sort
    public static List<Integer> mergeSort(List<Integer> arr) {
        if (arr.size() > 1) {
            int mid = arr.size() / 2;
            List<Integer> L = new ArrayList<>(arr.subList(0, mid));
            List<Integer> R = new ArrayList<>(arr.subList(mid, arr.size()));

            mergeSort(L);
            mergeSort(R);

            int i = 0, j = 0, k = 0;
            while (i < L.size() && j < R.size()) {
                if (L.get(i) < R.get(j)) {
                    arr.set(k++, L.get(i++));
                } else {
                    arr.set(k++, R.get(j++));
                }
            }
            while (i < L.size()) {
                arr.set(k++, L.get(i++));
            }
            while (j < R.size()) {
                arr.set(k++, R.get(j++));
            }
        }
        return arr;
    }

    // Função para ler números de um arquivo
    public static List<Integer> readNumbers(String fileName) throws IOException {
        List<String> lines = Files.readAllLines(Paths.get(fileName));
        List<Integer> numbers = new ArrayList<>();
        for (String line : lines) {
            numbers.add(Integer.parseInt(line.trim()));
        }
        return numbers;
    }

    // Função para escrever números em um arquivo
    public static void writeNumbers(String fileName, List<Integer> numbers) throws IOException {
        BufferedWriter writer = new BufferedWriter(new FileWriter(fileName));
        for (Integer number : numbers) {
            writer.write(number + "\n");
        }
        writer.close();
    }
}

File: Atividade 3/MergeSort.py
import time
import tracemalloc

# Função para ordenar uma lista de números usando Merge Sort
def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]

        merge_sort(L)
        merge_sort(R)

        i = j = k = 0

        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

    return arr

# Função para ler números de um arquivo
def read_numbers(file_name):
    with open(file_name, 'r') as file:
        numbers = [int(line.strip()) for line in file]
    return numbers

# Função para escrever números em um arquivo
def write_numbers(file_name, numbers):
    with open(file_name, 'w') as file:
        for number in numbers:
            file.write(f"{number}\n")

def main():
    # Lê os números do arquivo de entrada
    numbers = read_numbers('D:/Faculdade/Analise/An-lise-de-Desempenho/Atividade 3/arq (1).txt')

    # Inicia a medição de memória
    tracemalloc.start()
    # Inicia a medição de tempo
    start_time = time.time()
    
    # Ordena os números
    sorted_numbers = merge_sort(numbers)
    
    # Termina a medição de tempo
    end_time = time.time()
    # Termina a medição de memória
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    
    # Escreve os números ordenados no arquivo de saída
    write_numbers('arq-saida.txt', sorted_numbers)

    # Imprime o tempo de execução e a memória utilizada
    print(f"Tempo de execução: {(end_time - start_time) * 1000:.2f} ms")
    print(f"Memória utilizada: {peak / 1024:.2f} KB")

if __name__ == "__main__":
    main()
